### PR TITLE
N3 create crud controller

### PR DESCRIPTION
**Pull request by dmkofgl**

### CODE REVIEW

File: src/main/java/dl/news/portal/Application.java [hunk 1]
```diff
@@ -1,11 +1,19 @@
...
+import org.springframework.context.annotation.Bean;
 
 @SpringBootApplication
 public class Application {
     public static void main(String[] args) {
         SpringApplication.run(Application.class, args);
     }
+
+    @Bean
+    public Module hibernate5Module() {
```
Comment (avinahradau) [inline src/main/java/dl/news/portal/Application.java]: Why do you need here? You use spring boot I think it's redundant

File: src/main/java/dl/news/portal/domain/dto/DtoTransform.java [hunk 2]
```diff
@@ -0,0 +1,6 @@
+package dl.news.portal.domain.dto;
+
+@FunctionalInterface
+public interface DtoTransform<T> {
```
Comment (avinahradau) [inline src/main/java/dl/news/portal/domain/dto/DtoTransform.java]: I think this interface is redundant

File: src/main/java/dl/news/portal/domain/dto/NewsDto.java [hunk 3]
```diff
@@ -1,29 +1,53 @@
...
 
     @Override
     public void transfer(News receiver) {
         title.ifPresent(receiver::setTitle);
         content.ifPresent(receiver::setContent);
-        author.ifPresent(receiver::setAuthor);
+    }
+
+    @Override
+    public News transform() {
```
Comment (avinahradau) [inline src/main/java/dl/news/portal/domain/dto/NewsDto.java]: You can just use method static method  `of`

File: src/main/java/dl/news/portal/domain/dto/UserUpdateDto.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
+package dl.news.portal.domain.dto;
+
+import dl.news.portal.domain.entity.User;
+
+import javax.validation.constraints.Email;
+import javax.validation.constraints.Size;
+import java.util.Optional;
+
+public class UserUpdateDto implements DtoTransfer<User> {
```
Comment (avinahradau) [inline src/main/java/dl/news/portal/domain/dto/UserUpdateDto.java]: You can create just `UserDto` withou creating and update dtos. Please refactor it

**Review by avinahradau — CHANGES_REQUESTED** [review]

File: src/main/java/dl/news/portal/Application.java (hunk 1 above)
Comment (dmkofgl) [reply src/main/java/dl/news/portal/Application.java]: Ok. It's removed.

File: src/main/java/dl/news/portal/domain/dto/UserUpdateDto.java (hunk 4 above)
Comment (dmkofgl) [reply src/main/java/dl/news/portal/domain/dto/UserUpdateDto.java]: Done

File: src/main/java/dl/news/portal/domain/dto/NewsDto.java (hunk 3 above)
Comment (dmkofgl) [reply src/main/java/dl/news/portal/domain/dto/NewsDto.java]: Ok

File: src/main/java/dl/news/portal/domain/dto/DtoTransform.java (hunk 2 above)
Comment (dmkofgl) [reply src/main/java/dl/news/portal/domain/dto/DtoTransform.java]: Ok. I agree.

File: src/main/java/dl/news/portal/domain/dto/NewsDto.java [hunk 5]
```diff
@@ -1,29 +1,51 @@
...
     }
 
-    @Override
     public void transfer(News receiver) {
         title.ifPresent(receiver::setTitle);
         content.ifPresent(receiver::setContent);
-        author.ifPresent(receiver::setAuthor);
+    }
+
+    public static News of(NewsDto newsDto) {
```
Comment (avinahradau) [inline src/main/java/dl/news/portal/domain/dto/NewsDto.java]: You don't need `transfer` method anymore. You just need all and that all

File: src/main/java/dl/news/portal/domain/dto/UserDto.java [hunk 6]
```diff
@@ -2,24 +2,54 @@
...
+
+    public UserDto(String username, String email, String password) {
+        this.username = Optional.ofNullable(username);
+        this.email = Optional.ofNullable(email);
+        this.password = Optional.ofNullable(password);
     }
 
     public UserDto() {
-        this(null, null);
+        this(null, null, null);
```
Comment (avinahradau) [inline src/main/java/dl/news/portal/domain/dto/UserDto.java]: absolutely  redundant constructor. Remove it

File: src/main/java/dl/news/portal/domain/dto/UserDto.java [hunk 7]
```diff
@@ -2,24 +2,54 @@
...
-        optionalUsername.ifPresent(receiver::setUsername);
-        optionalEmail.ifPresent(receiver::setEmail);
+        username.ifPresent(receiver::setUsername);
+        email.ifPresent(receiver::setEmail);
+        password.ifPresent(receiver::setPassword);
+    }
+
+    public static User of(UserDto userDto) {
+        User user = new User();
+        userDto.transfer(user);
```
Comment (avinahradau) [inline src/main/java/dl/news/portal/domain/dto/UserDto.java]: as explained above

File: src/main/java/dl/news/portal/web/controller/NewsController.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+        return new ResponseEntity<>(HttpStatus.CREATED);
+    }
+
+    @PatchMapping("/{id}")
+    public ResponseEntity<HttpStatus> updateNews(@PathVariable Long id, @Valid @RequestBody NewsDto newsDto) {
+        newsService.updateNews(id, newsDto);
+        return new ResponseEntity<>(HttpStatus.OK);
+    }
+
+    @DeleteMapping("/{id}")
```
Comment (avinahradau) [inline src/main/java/dl/news/portal/web/controller/NewsController.java]: move `"/{id}"` to the constant

File: src/main/java/dl/news/portal/web/controller/RestExceptionHandler.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    @ExceptionHandler({DeniedOperationException.class})
+    public ResponseEntity<Resource<String>> handleDeniedOperationException(DeniedOperationException ex) {
+        Resource<String> resource = new Resource<>(ex.getMessage());
+        return new ResponseEntity<>(resource, HttpStatus.UNPROCESSABLE_ENTITY);
+    }
+
+    @ExceptionHandler({EntityNotFoundException.class})
+    public ResponseEntity<Resource<String>> handleEntityNotFoundException(EntityNotFoundException ex) {
```
Comment (avinahradau) [inline src/main/java/dl/news/portal/web/controller/RestExceptionHandler.java]: What is difference with EntityNotFound and EntityNotExists exceptions ?

File: src/test/java/dl/news/portal/domain/service/NewsServiceTest.java [hunk 10]
```diff
@@ -85,4 +87,11 @@ public void findByUpdatedDateBetween() {
         assertNotEquals(allNews.size(), news.size());
     }
 
+    @Test
+    public void findPageByAuthor() {
```
Comment (avinahradau) [inline src/test/java/dl/news/portal/domain/service/NewsServiceTest.java]: how this test work? What it tests ?

File: src/test/java/dl/news/portal/web/controller/NewsControllerTest.java [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+        mapper.setSerializationInclusion(JsonInclude.Include.NON_NULL);
+        String body = mapper.writeValueAsString(NewsDto.of(news));
+        mockMvc.perform(post("/news").content(body).contentType(MediaType.APPLICATION_JSON)).andExpect(status().isCreated());
+        news.setTitle(null);
+        body = mapper.writeValueAsString(NewsDto.of(news));
+        mockMvc.perform(post("/news").content(body).contentType(MediaType.APPLICATION_JSON)).andExpect(status().isUnprocessableEntity());
+    }
+
+    @Test
+    public void addNewsNegative() throws Exception {
```
Comment (avinahradau) [inline src/test/java/dl/news/portal/web/controller/NewsControllerTest.java]: You should add negative tests, for example some tests taht can returns nothing or exceptions. and use naming convention like : 

> whatTestDo_whatWehave_whatShouldReturn

example: 
`addNews_whenDtoIsValid_shouldReturnCreated`

Also your tests you write a lot of checks, you always should do one action in one test. if it create news, you just should check that news is created and that all. Use different test for user test case

File: src/test/java/dl/news/portal/web/controller/UserControllerTest.java [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+    public void getUserById() throws Exception {
+        User user = new User();
+        user.setId(1L);
+        user.setUsername("testUsername");
+        user.setEmail("testEmail@mail.com");
+
+        Mockito.when(userService.findUserById(1L)).thenReturn(Optional.of(user));
+
+        mockMvc.perform(get("/users/1")).andExpect(status().isOk());
+        mockMvc.perform(get("/users/2")).andExpect(status().isNotFound());
```
Comment (avinahradau) [inline src/test/java/dl/news/portal/web/controller/UserControllerTest.java]: Use separated test method for user not found

**Review by avinahradau — CHANGES_REQUESTED** [review]

File: src/main/java/dl/news/portal/web/controller/NewsController.java (hunk 8 above)
Comment (dmkofgl) [reply src/main/java/dl/news/portal/web/controller/NewsController.java]: Done

File: src/test/java/dl/news/portal/web/controller/UserControllerTest.java (hunk 12 above)
Comment (dmkofgl) [reply src/test/java/dl/news/portal/web/controller/UserControllerTest.java]: Done

File: src/test/java/dl/news/portal/web/controller/NewsControllerTest.java (hunk 11 above)
Comment (dmkofgl) [reply src/test/java/dl/news/portal/web/controller/NewsControllerTest.java]: Use this name convention for all tests or only for controllers?

File: src/test/java/dl/news/portal/domain/service/NewsServiceTest.java (hunk 10 above)
Comment (dmkofgl) [reply src/test/java/dl/news/portal/domain/service/NewsServiceTest.java]: Nothing. This test and service method redundant and removed in searching branch

File: src/main/java/dl/news/portal/domain/dto/UserDto.java (hunk 6 above)
Comment (dmkofgl) [reply src/main/java/dl/news/portal/domain/dto/UserDto.java]: I see

File: src/main/java/dl/news/portal/domain/response/PageResponse.java [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+    private void addPageableLinks() {
+        int firstPageNumber = 0;
+        int lastPageNumber = page.getTotalPages() <= 0 ? 0 : page.getTotalPages() - 1;
+        addPageNumberLinkWithRel(firstPageNumber, Link.REL_FIRST);
+        addPageNumberLinkWithRel(lastPageNumber, Link.REL_LAST);
+        if (page.hasNext()) {
+            int nextPageNumber = page.nextPageable().getPageNumber();
+            addPageNumberLinkWithRel(nextPageNumber, Link.REL_NEXT);
+        }
+        if (page.hasPrevious()) {
```
Comment (avinahradau) [inline src/main/java/dl/news/portal/domain/response/PageResponse.java:29]: Just suggestion, add empty line before if()

File: src/main/java/dl/news/portal/domain/service/impl/NewsServiceImpl.java [hunk 14]
```diff
@@ -35,23 +35,19 @@
     }
 
     @Override
-    public void saveNews(News news) {
+    public void createNews(NewsDto dto) {
+        News news = new News();
+        transferDto(news, dto);
```
Comment (avinahradau) [inline src/main/java/dl/news/portal/domain/service/impl/NewsServiceImpl.java]: rename 

>  transferDto

to `transformDto`

**Review by avinahradau — APPROVED** [review]

File: src/test/java/dl/news/portal/domain/service/NewsServiceTest.java [hunk 15]
```diff
@@ -39,29 +39,23 @@ public void findByTitle() {
...
 
     @Test
-    public void addNews() {
-        User user = new User();
-        user.setId(1L);
-        News news = new News();
-        news.setTitle("test title");
-        news.setContent("content");
-        news.setAuthor(user);
+    public void addNews_whenNewsDtoIsvalid_shouldChangeSize() {
```
Comment (avinahradau) [inline src/test/java/dl/news/portal/domain/service/NewsServiceTest.java]: Punctuation error `addNews_whenNewsDtoIsValid_shouldChangeSize()`

File: src/test/java/dl/news/portal/web/controller/NewsControllerTest.java [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+        User user = new User();
+        user.setId(2L);
+        News news = new News();
+        news.setId(1L);
+        news.setContent("test");
+        news.setTitle("test");
+        news.setAuthor(user);
+
+        Mockito.when(newsService.findNewsById(1L)).thenReturn(Optional.of(news));
+        mockMvc.perform(get(path)).andExpect(status().isOk());
```
Comment (avinahradau) [inline src/test/java/dl/news/portal/web/controller/NewsControllerTest.java]: Add what else test expect for example fields

File: src/test/java/dl/news/portal/domain/service/NewsServiceTest.java (hunk 15 above)
Comment (dmkofgl) [reply src/test/java/dl/news/portal/domain/service/NewsServiceTest.java]: I see

File: src/main/java/dl/news/portal/domain/service/impl/NewsServiceImpl.java (hunk 14 above)
Comment (dmkofgl) [reply src/main/java/dl/news/portal/domain/service/impl/NewsServiceImpl.java]: Done

File: src/test/java/dl/news/portal/web/controller/NewsControllerTest.java (hunk 16 above)
Comment (dmkofgl) [reply src/test/java/dl/news/portal/web/controller/NewsControllerTest.java]: Done

**Pull request merged**